pypeit/images/rawimage.py

```diff
@@ -19,399 +19,701 @@
+        hdu (`astropy.io.fits.HDUList`_):
+            The full list of HDUs provided by :attr:`filename`.
+        exptime (:obj:`float`):
+            Frame exposure time in seconds.
+        rawdatasec_img (`numpy.ndarray`_):
+            The original, not trimmed or reoriented, image identifying which
+            amplifier was used to read each section of the raw image.
+        oscansec_img (`numpy.ndarray`_):
+            The original, not trimmed or reoriented, image identifying the
+            overscan regions in the raw image read by each amplifier.
+        headarr (:obj:`list`):
+            A list of `astropy.io.fits.Header`_ objects with the headers for all
+            extensions in :attr:`hdu`.
         image (`numpy.ndarray`_):
+            The processed image.  This starts as identical to :attr:`rawimage`
+            and then altered by the processing steps; see :func:`process`.
+        ronoise (:obj:`list`):
+            The readnoise (in e-/ADU) for each of the detector amplifiers.
+        par (:class:`~pypeit.par.pypeitpar.ProcessImagesPar`):
+            Parameters that dictate the processing of the images.
+        ivar (`numpy.ndarray`_):
+            The inverse variance of :attr:`image`, the processed image.
+        rn2img (`numpy.ndarray`_):
+            The readnoise variance image.
+        steps (:obj:`dict`):
+            Dictionary containing a set of booleans that track the processing
+            steps that have been performed.
+        datasec_img (`numpy.ndarray`_):
+            Image identifying which amplifier was used to read each section of
+            the *processed* image.
+        spat_flexure_shift (:obj:`float`):
+            The spatial flexure shift in pixels, if calculated
     """
     def __init__(self, ifile, spectrograph, det):
 
         # Required parameters
         self.filename = ifile
-        self.spectrograph = spectrograph  # One for convenience
+        self.spectrograph = spectrograph
         self.det = det
 
-        # Load
         # Load the raw image and the other items of interest
         self.detector, self.rawimage, self.hdu, self.exptime, self.rawdatasec_img, \
-            self.oscansec_img = self.spectrograph.get_rawimage(self.filename, self.det)
+                self.oscansec_img = self.spectrograph.get_rawimage(self.filename, self.det)
 
         # Grab items from rawImage (for convenience and for processing)
         #   Could just keep rawImage in the object, if preferred
         self.headarr = deepcopy(self.spectrograph.get_headarr(self.hdu))
 
         # Key attributes
-        self.rawimage = self.rawimage.copy()
         self.image = self.rawimage.copy()
         self.datasec_img = self.rawdatasec_img.copy()
-        self.ronoise = self.detector['ronoise']
+        # NOTE: Prevent estimate_readnoise() from altering self.detector using
+        # deepcopy
+        self.ronoise = deepcopy(self.detector['ronoise'])
```

Comment:
Why the deep copy here? Are we modifying the ronoise values?

Reply:
Yes, these can be modified using the estimate_readnoise function, and users can now select this approach using the emprical_rn parameter.